apps/emqx_ft/src/emqx_ft.erl

```diff
@@ -271,51 +222,92 @@ on_segment(#{packet_id := PacketId}, Msg, Transfer, Offset, Checksum) ->
     }),
     Segment = {Offset, Msg#message.payload},
     %% Currently synchronous.
-    %% If we want to make it async, we need to use `emqx_ft_async_reply`,
+    %% If we want to make it async, we need to use `with_responder`,
     %% like in `on_fin`.
     ?ACK_AND_PUBLISH(store_segment(Transfer, Segment)).
 
-on_fin(#{packet_id := PacketId} = TopicReplyData, Msg, Transfer, FinalSize, FinalChecksum) ->
+on_fin(
+    #{packet_id := PacketId, mode := Mode} = TopicReplyData, Msg, Transfer, FinalSize, FinalChecksum
+) ->
     ?tp(info, "file_transfer_fin", #{
         mqtt_msg => Msg,
         packet_id => PacketId,
         transfer => Transfer,
         final_size => FinalSize,
         checksum => FinalChecksum
     }),
-    %% TODO: handle checksum? Do we need it?
-    with_new_packet(
-        TopicReplyData,
-        PacketId,
-        fun() ->
-            case assemble(Transfer, FinalSize, FinalChecksum) of
-                ok ->
-                    ?ACK_AND_PUBLISH(ok);
-                %% Assembling started, packet will be acked/replied by monitor or timeout
-                {async, Pid} ->
-                    register_async_worker(Pid, TopicReplyData);
-                {error, _} = Error ->
-                    ?ACK_AND_PUBLISH(Error)
-            end
+    FinPacketKey = {self(), PacketId},
+    Callback = fun(Result) ->
+        on_complete("assemble", TopicReplyData, FinPacketKey, Transfer, Result)
+    end,
+    with_responder(FinPacketKey, Callback, emqx_ft_conf:assemble_timeout(), fun() ->
+        case assemble(Transfer, FinalSize, FinalChecksum) of
+            %% Assembling completed, ack through the responder right away
+            ok ->
+                emqx_ft_responder:ack(FinPacketKey, ok),
+                ?DELAY_ACK;
```

Comment:
Newbie question: is `?DELAY_ACK` intentional here, since it was `?ACK_AND_PUBLISH(ok)` before?

Reply:
Yes; the ack is done via the `Callback`.